replication: reset stale http pooled connections

abraithwaite opened the pull request:
I don't have experience writing C++, but this is generally what I think needs to be happening.  I'd love some assistance to get this to build and eventually merged mainline.  Or if one of the contributors wants to just take it, that would be much appreciated as well!

Related to:
https://github.com/yandex/ClickHouse/issues/5287

According to the Poco docs, we should be resetting the connection when
sendRequest or readResponse throws.

https://pocoproject.org/docs/Poco.Net.HTTPClientSession.html#21613

I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Category:
- Bug Fix

IvanKush commented on dbms/src/IO/HTTPCommon.cpp:
```diff
@@ -196,7 +197,18 @@ PooledHTTPSessionPtr makePooledHTTPSession(const Poco::URI & uri, const Connecti
-    auto istr = &session.receiveResponse(response);
+    try
+    {
+        auto istr = &session.receiveResponse(response);
```
istr here is local variable. 

I think you may
```
std::istream* istr = nullptr;
try {
   istr = 
}
catch() 
{
 ... your code
}

if (istr == nullptr) {
   LOG_ message
   throw Exception
}

```

IvanKush commented on dbms/src/IO/HTTPCommon.cpp:
```diff
@@ -196,7 +197,18 @@ PooledHTTPSessionPtr makePooledHTTPSession(const Poco::URI & uri, const Connecti
+    {
+        log = &Logger::get("IOHTTP");
+        LOG_DEBUG(log, "reason: " << e.displayText());
+        &session->reset();
```
simply `session.reset();`

IvanKush commented on dbms/src/IO/HTTPCommon.cpp:
```diff
@@ -196,7 +197,18 @@ PooledHTTPSessionPtr makePooledHTTPSession(const Poco::URI & uri, const Connecti
+        LOG_DEBUG(log, "reason: " << e.displayText());
+        &session->reset();
+        // Retry once
+        istr = &session.receiveResponse(request);
```
I think, the logic of retry is better to make in loop.
Maybe, add parameter `Retries` in `Settings.h` and loop `Retries` time. If after `Retries` count still we have exceptions, so log the error, throw exception, etc.

abraithwaite replied:
Good call.  I'm going to test this in our deployment before I dive into configuration, but should definitely be configurable.  I'll look around the codebase for other examples of retries and try to match them, unless a contributor beats me to it.

Commit:
replication: reset stale http pooled connections

Related to:
https://github.com/yandex/ClickHouse/issues/5287

According to the Poco docs, we should be resetting the connection when
sendRequest or readResponse throws.

https://pocoproject.org/docs/Poco.Net.HTTPClientSession.html#21613

nvartolomei commented:
irc send resets connection under the hood if last request errored

nvartolomei commented:
https://github.com/yandex/ClickHouse/pull/4837#issuecomment-477930639

abraithwaite commented:
Thanks for the pointers!  Reading that code now, but I'm confused as to why the Poco docs still say that resetting the connection is required when issues are encountered.

Basically trying to hunt down why we'd be connecting to a stale IP address after the host changes what it's pointing to (#5287)

abraithwaite commented:
Digging into the Poco codebase, that's some pretty complex code.  I've got strong reason to believe that the connections are not being reinitialized though.

One thing I find interesting however is that `connected()` calls `socket.initialized()` which only checks `_sockfd != POCO_INVALID_SOCKET`.  `_sockfd` only gets set to `POCO_INVALID_SOCKET` when `close()` is called on the socket, meaning that without an explicit close, `connected()` will always be true.  In our case, we've got a broken/stale HTTP connection and I don't see anywhere `close()` would be called proactively in `sendRequest` or `receiveResponse`.

I'm going to look into `keepAlive` next since that seems to be the sole factor controlling `mustReconnect` in that client.

Given what I've seen so far though, I'm thinking `reset()` still needs to be called after net exceptions.

abraithwaite commented:
I'm an idiot.  `close()` is called in the except blocks for both `sendRequest()` and `receiveResponse()`.  Dead end...

Enmk commented:
> I'm an idiot. `close()` is called in the except blocks for both `sendRequest()` and `receiveResponse()` in Poco libraries. Dead end...

Don't blame yourself, it is complicated. In fact, I think that your fix might work, regardless of catching an exception in `HTTPClientSession::sendRequest()`. 

(please help me do a theory check) What if connection was established at first, but then during download you've got connectivity issues: 
1. response body is partially received and stream-object is returned to the caller of `DB::receiveResponse()`
2. someone reads from that stream, and during that, either side of session experience connectivity issues, causing a bad read. Exception is thrown, but the user of stream has no means to reset the session, leaving socket in a bad state.
3. Next request on this session would fail (since the socket is in the bad state), but `HTTPClientSession::sendRequest()` will close the socket in except-block.
4. Next-after-that request re-opens a socket and has a chance to succeed.

Without a fix, user of this HTTP-session would experience first and second requests to fail, but third may succeed.

With your fix, user of this HTTP-session would experience first request to fail, but the second one may succeed.

I think there might be a simpler alternative to this fix, you could just reset the session before returning it from the pool in dbms/src/IO/HTTPCommon.cpp  `HTTPSessionPool::getSession`:
```
            if (session->netException())
            {
                session->reset();
            }
```

abraithwaite commented:
@nvartolomei was right, poco handles it correctly.

I believe this is what's happening now:

https://github.com/yandex/ClickHouse/issues/5287#issuecomment-495071521

@Enmk maybe you could verify?

Enmk commented:
> @nvartolomei was right, poco handles it correctly.
> 
> I believe this is what's happening now:
> 
> [#5287 (comment)](https://github.com/yandex/ClickHouse/issues/5287#issuecomment-495071521)
> 
> @Enmk WDYT about my current theory ^

Sounds legit, there are two options now:
* Key by IP. I believe that would cause dead sessions to pile up once the host changes IP.
* Key by host (as it is now), but reset the host once in a while. So basically, two birds, one stone:
(almost as with https://github.com/yandex/ClickHouse/pull/5381#issuecomment-495066358, but also reset the host):
```
if (session->networkException())
{
    session->reset();
    session->setHost(DNSResolver::instance().resolveHost(host).toString());
}
```

abraithwaite commented:
> I believe that we still need to resolve host to IP to be consistent with rest of the codebase.

Respectfully, I think that's the wrong choice here.  This is a pooled connection, meaning we're already doing our best to reuse resources and not do unnecessary DNS lookups.

This is the only place `setHost` is called with an IP not the host itself.  One call even has a comment to that regard:

```
dbms/src/IO/HTTPCommon.cpp
83:        session->setHost(DNSResolver::instance().resolveHost(host).toString());

dbms/src/IO/WriteBufferFromHTTP.cpp
15:    request.setHost(uri.getHost());

dbms/src/IO/ReadWriteBufferFromHTTP.h
60:            request.setHost(uri.getHost()); // use original, not resolved host name in header

dbms/src/Storages/MergeTree/DataPartsExchange.cpp
173:    uri.setHost(host);

dbms/src/Common/XDBCBridgeHelper.h
88:        ping_url.setHost(bridge_host);
191:        uri.setHost(ping_url.getHost());
```

Enmk commented:
>  This is a pooled connection, meaning we're already doing our best to reuse resources and not do unnecessary DNS lookups.

well, maybe... I do not know for sure what was the intention behind this. @alesapin could you please clarify ?

abraithwaite commented:
Looking at the git history, it seems like it was an attempt to use the internal caching resolver over just resolving the hostname.  Following it backwards in the git history, it seems quite old.

Maybe we should move the conversation back to #5287 just to keep everything in one place.

I really appreciate your help and insight!  Thanks for taking a look at it! :-)

abraithwaite commented:
Actually, the last comment that line before `SYSTEM DROP DNS CACHE` had was to cache DNS until restart:

https://github.com/yandex/ClickHouse/commit/48f5d8f1a3e0f980e1d559d0a5cd4b9ebbe29220
